From 4eb5b86acdcf9585b5937249462c8ee7a8f773ff Mon Sep 17 00:00:00 2001
From: gbhakteX <gurudattax.bhakte@intel.com>
Date: Fri, 18 Jan 2013 14:31:37 +0530
Subject: Klocwork Fix required in Stagefright engine.

BZ: 81354

Klocwork issue #2817: Resource acquired to 'new(sizeof(FileSource))->mBuffer.mLock.mMutex'
at line 147 may be lost here. Also there are 3 similar errors on line(s) 193, 196.
Fix: 'source' is cleared before returning NULL, to avoid resource leak.
[DataSource.cpp | CreateFromURI() | line 193].

Klocwork issue #5255: Resource acquired to 'new(sizeof(FileSource))->mBuffer.mLock.mMutex'
at line 222 may be lost here. Also there is one similar error on line 310.
Fix: 'source' (only if it is FileSource) is cleared before returning from function
fetchFile(), to avoid resource leak.
[LiveSession.cpp | fetchFile() | line 298, 313].

Category: AOSP Improvement
Domain: AUDIO_MEDIA-Audio Framework
Origin: Internal
Upstream-Candidate: no, proprietary

Change-Id: I899a1f309bcca4f1afdcf36737969a33c33368a3
Orig-Change-Id: I73f625a096658d7aa85294335bc8e694198cfbbc
Signed-off-by: gbhakteX <gurudattax.bhakte@intel.com>
Signed-off-by: Arulselvan M <arulselvan.m@intel.com>
---
 media/libstagefright/DataSource.cpp           |    3 +++
 media/libstagefright/httplive/LiveSession.cpp |    6 ++++++
 2 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/DataSource.cpp b/media/libstagefright/DataSource.cpp
index 53802d4..737bb5f 100644
--- a/media/libstagefright/DataSource.cpp
+++ b/media/libstagefright/DataSource.cpp
@@ -239,6 +239,9 @@ sp<DataSource> DataSource::CreateFromURI(
     }
 
     if (source == NULL || source->initCheck() != OK) {
+        if (source != NULL) {
+            source.clear();
+        }
         return NULL;
     }
 
diff --git a/media/libstagefright/httplive/LiveSession.cpp b/media/libstagefright/httplive/LiveSession.cpp
index bd12ddc..d0943d8 100644
--- a/media/libstagefright/httplive/LiveSession.cpp
+++ b/media/libstagefright/httplive/LiveSession.cpp
@@ -588,6 +588,9 @@ status_t LiveSession::fetchFile(
                 maxBytesToRead);
 
         if (n < 0) {
+            if ((!strncasecmp(url, "file://", 7)) && (source != NULL)) {
+                source.clear();
+            }
             return n;
         }
 
@@ -600,6 +603,9 @@ status_t LiveSession::fetchFile(
 
     *out = buffer;
 
+    if ((!strncasecmp(url, "file://", 7)) && (source != NULL)) {
+        source.clear();
+    }
     return OK;
 }
 
-- 
1.7.4.1

